File: lib/template-engine.ts
/**
 * Template Engine for Campaign Messages
 * Handles variable replacement and spintax processing
 */

interface Lead {
  first_name?: string
  last_name?: string
  full_name?: string
  company?: string
  position?: string
  headline?: string
  location?: string
  email?: string
  linkedin_url?: string
  [key: string]: any
}

interface TemplateVariables {
  [key: string]: string | undefined
}

/**
 * Replace template variables with lead data
 * Supports: {{firstName}}, {{lastName}}, {{company}}, {{position}}, etc.
 */
export function replaceVariables(template: string, lead: Lead): string {
  if (!template) return ''
  
  // Build variable map
  const variables: TemplateVariables = {
    firstName: lead.first_name || lead.full_name?.split(' ')[0] || '',
    lastName: lead.last_name || lead.full_name?.split(' ').slice(1).join(' ') || '',
    fullName: lead.full_name || `${lead.first_name || ''} ${lead.last_name || ''}`.trim(),
    company: lead.company || '',
    position: lead.position || lead.headline || '',
    headline: lead.headline || lead.position || '',
    location: lead.location || '',
    email: lead.email || '',
    linkedinUrl: lead.linkedin_url || '',
    aiIcebreaker: lead.ai_icebreaker || ''
  }
  
  // Replace all variables
  let result = template
  
  Object.entries(variables).forEach(([key, value]) => {
    const regex = new RegExp(`{{${key}}}`, 'gi')
    result = result.replace(regex, value || '')
  })
  
  // Clean up any remaining unreplaced variables (replace with empty string)
  result = result.replace(/{{[^}]+}}/g, '')
  
  return result.trim()
}

/**
 * Process spintax: {Option1|Option2|Option3}
 * Randomly selects one option from pipe-separated list
 */
export function processSpintax(text: string): string {
  if (!text) return ''
  
  const spintaxRegex = /{([^}]+)}/g
  
  let result = text
  let match: RegExpExecArray | null
  
  // Find all spintax patterns
  while ((match = spintaxRegex.exec(text)) !== null) {
    const fullMatch = match[0] // {Option1|Option2|Option3}
    const options = match[1].split('|').map(opt => opt.trim()) // ['Option1', 'Option2', 'Option3']
    
    // Randomly select one option
    const selectedOption = options[Math.floor(Math.random() * options.length)]
    
    // Replace the spintax with selected option
    result = result.replace(fullMatch, selectedOption)
  }
  
  return result
}

/**
 * Validate template syntax
 * Returns array of errors if invalid
 */
export function validateTemplate(template: string): string[] {
  const errors: string[] = []
  
  if (!template || template.trim().length === 0) {
    errors.push('Template cannot be empty')
    return errors
  }
  
  // Check for unclosed variable braces
  const openBraces = (template.match(/{{/g) || []).length
  const closeBraces = (template.match(/}}/g) || []).length
  
  if (openBraces !== closeBraces) {
    errors.push('Unclosed variable braces detected ({{ or }})')
  }
  
  // Check for unclosed spintax braces
  const openSpintax = (template.match(/{(?!{)/g) || []).length
  const closeSpintax = (template.match(/}(?!})/g) || []).length
  
  if (openSpintax !== closeSpintax) {
    errors.push('Unclosed spintax braces detected ({ or })')
  }
  
  // Check for nested braces (not supported)
  if (/{{[^}]*{{/.test(template) || /{[^}]*{[^{]/.test(template)) {
    errors.push('Nested braces are not supported')
  }
  
  // Warn about unknown variables
  const knownVariables = [
    'firstName', 'lastName', 'fullName', 'company', 'position',
    'headline', 'location', 'email', 'linkedinUrl', 'aiIcebreaker'
  ]
  
  const variablePattern = /{{([^}]+)}}/g
  let match: RegExpExecArray | null
  
  while ((match = variablePattern.exec(template)) !== null) {
    const varName = match[1].trim()
    if (!knownVariables.includes(varName)) {
      errors.push(`Unknown variable: {{${varName}}}. Available: ${knownVariables.join(', ')}`)
    }
  }
  
  return errors
}

/**
 * Get list of variables used in template
 */
export function extractVariables(template: string): string[] {
  const variables: string[] = []
  const variablePattern = /{{([^}]+)}}/g
  let match: RegExpExecArray | null
  
  while ((match = variablePattern.exec(template)) !== null) {
    const varName = match[1].trim()
    if (!variables.includes(varName)) {
      variables.push(varName)
    }
  }
  
  return variables
}

/**
 * Check if lead has all required data for template
 */
export function hasRequiredData(template: string, lead: Lead): { 
  valid: boolean
  missing: string[] 
} {
  const variables = extractVariables(template)
  const missing: string[] = []
  
  const leadData: TemplateVariables = {
    firstName: lead.first_name || lead.full_name?.split(' ')[0],
    lastName: lead.last_name || lead.full_name?.split(' ').slice(1).join(' '),
    fullName: lead.full_name,
    company: lead.company,
    position: lead.position || lead.headline,
    headline: lead.headline || lead.position,
    location: lead.location,
    email: lead.email,
    linkedinUrl: lead.linkedin_url,
    aiIcebreaker: lead.ai_icebreaker
  }
  
  variables.forEach(varName => {
    if (!leadData[varName] || leadData[varName]?.trim().length === 0) {
      missing.push(varName)
    }
  })
  
  return {
    valid: missing.length === 0,
    missing
  }
}

/**
 * Generate preview of template with lead data
 */
export function previewTemplate(template: string, lead: Lead): {
  preview: string
  warnings: string[]
} {
  const warnings: string[] = []
  
  // Validate template first
  const errors = validateTemplate(template)
  if (errors.length > 0) {
    return {
      preview: template,
      warnings: errors
    }
  }
  
  // Check for missing data
  const { missing } = hasRequiredData(template, lead)
  if (missing.length > 0) {
    warnings.push(`Missing data for: ${missing.join(', ')}. These will be replaced with empty strings.`)
  }
  
  // Process template
  let preview = replaceVariables(template, lead)
  preview = processSpintax(preview)
  
  return {
    preview,
    warnings
  }
}

/**
 * Process complete message (variables + spintax)
 */
export function processTemplate(template: string, lead: Lead): string {
  // First replace variables
  let processed = replaceVariables(template, lead)
  
  // Then process spintax
  processed = processSpintax(processed)
  
  return processed
}

/**
 * Check character limits for different message types
 */
export function checkCharacterLimit(
  text: string, 
  type: 'connection_note' | 'message' | 'inmail'
): { 
  valid: boolean
  length: number
  limit: number
  overflow: number
} {
  const limits = {
    connection_note: 300,
    message: 8000,
    inmail: 1900 // InMail body limit
  }
  
  const limit = limits[type]
  const length = text.length
  
  return {
    valid: length <= limit,
    length,
    limit,
    overflow: Math.max(0, length - limit)
  }
}

/**
 * Example usage and testing
 */
export function testTemplateEngine() {
  const testLead: Lead = {
    first_name: 'John',
    last_name: 'Doe',
    full_name: 'John Doe',
    company: 'Acme Corp',
    position: 'VP of Sales',
    location: 'San Francisco, CA'
  }
  
  const templates = [
    "Hi {{firstName}}, I noticed you work at {{company}}. {Let's connect!|Would love to connect!}",
    "Hey {{fullName}}, saw your profile. {{position}} at {{company}} is impressive!",
    "{{firstName}}, {your background|your experience} in {{industry}} caught my attention",
    "Hi {{firstName}}, {{invalidVar}} test" // Should show warning
  ]
  
  console.log('=== Template Engine Test ===\n')
  
  templates.forEach((template, index) => {
    console.log(`Template ${index + 1}: ${template}`)
    
    const validation = validateTemplate(template)
    if (validation.length > 0) {
      console.log('  Validation Errors:', validation)
    }
    
    const { preview, warnings } = previewTemplate(template, testLead)
    console.log('  Preview:', preview)
    
    if (warnings.length > 0) {
      console.log('  Warnings:', warnings)
    }
    
    console.log('')
  })
}

File: backend/internal/service/template_engine.go
package service

import (
	"fmt"
	"math/rand"
	"regexp"
	"strings"
)

// ── Character Limits (LinkedIn enforced) ────────────────────────────────────

var characterLimits = map[string]int{
	"connection_note": 300,
	"message":         8000,
	"inmail":          1900,
}

// CharacterLimitResult holds the result of a character limit check.
type CharacterLimitResult struct {
	Valid    bool     `json:"valid"`
	Length   int      `json:"length"`
	Limit    int      `json:"limit"`
	Overflow int      `json:"overflow"`
	Warnings []string `json:"warnings,omitempty"`
}

// CheckCharacterLimit validates message length against LinkedIn limits.
func CheckCharacterLimit(message, messageType string) CharacterLimitResult {
	limit, ok := characterLimits[messageType]
	if !ok {
		limit = 8000 // default to message limit
	}

	length := len(message)
	overflow := 0
	var warnings []string
	if length > limit {
		overflow = length - limit
		warnings = append(warnings, fmt.Sprintf("Message exceeds %s limit by %d characters", messageType, overflow))
	} else if float64(length) > float64(limit)*0.9 {
		warnings = append(warnings, fmt.Sprintf("Message is within 10%% of %s character limit (%d/%d)", messageType, length, limit))
	}

	return CharacterLimitResult{
		Valid:    length <= limit,
		Length:   length,
		Limit:    limit,
		Overflow: overflow,
		Warnings: warnings,
	}
}

// ── Template Variable Replacement ───────────────────────────────────────────

// TemplateVariables holds lead data for variable replacement.
type TemplateVariables struct {
	FirstName    string `json:"first_name"`
	LastName     string `json:"last_name"`
	FullName     string `json:"full_name"`
	Company      string `json:"company"`
	Position     string `json:"position"`
	Location     string `json:"location"`
	Headline     string `json:"headline"`
	Email        string `json:"email"`
	LinkedInURL  string `json:"linkedin_url"`
	AIIcebreaker string `json:"ai_icebreaker"`
}

// variableMap returns the lookup map for template variables.
func (v *TemplateVariables) variableMap() map[string]string {
	return map[string]string{
		"firstName":    v.FirstName,
		"lastName":     v.LastName,
		"fullName":     v.FullName,
		"name":         v.FirstName, // alias
		"company":      v.Company,
		"position":     v.Position,
		"jobTitle":     v.Position, // alias
		"location":     v.Location,
		"headline":     v.Headline,
		"email":        v.Email,
		"linkedinUrl":  v.LinkedInURL,
		"aiIcebreaker": v.AIIcebreaker,
	}
}

var templateVarRegex = regexp.MustCompile(`\{\{(\w+)\}\}`)

// ReplaceVariables substitutes {{variable}} placeholders with lead data.
func ReplaceVariables(template string, vars *TemplateVariables) string {
	if vars == nil {
		return template
	}

	varMap := vars.variableMap()

	return templateVarRegex.ReplaceAllStringFunc(template, func(match string) string {
		// Extract the variable name from {{varName}}
		key := match[2 : len(match)-2]
		if val, ok := varMap[key]; ok && val != "" {
			return val
		}
		return "" // strip unresolved variables (matches TS behavior)
	})
}

// ── Spintax Processing ──────────────────────────────────────────────────────

var spintaxRegex = regexp.MustCompile(`\{([^{}]+)\}`)

// ProcessSpintax resolves {option1|option2|option3} syntax by picking a random option.
func ProcessSpintax(text string) string {
	return spintaxRegex.ReplaceAllStringFunc(text, func(match string) string {
		// Remove outer braces
		inner := match[1 : len(match)-1]
		options := strings.Split(inner, "|")
		if len(options) == 0 {
			return match
		}
		// Trim whitespace from each option
		for i := range options {
			options[i] = strings.TrimSpace(options[i])
		}
		return options[rand.Intn(len(options))]
	})
}

// ── Combined Processing ─────────────────────────────────────────────────────

// ProcessTemplate applies variable replacement and spintax processing.
func ProcessTemplate(template string, vars *TemplateVariables) string {
	result := ReplaceVariables(template, vars)
	result = ProcessSpintax(result)
	return result
}

// PreviewTemplate processes a template for preview (same as ProcessTemplate).
func PreviewTemplate(template string, vars *TemplateVariables) string {
	return ProcessTemplate(template, vars)
}

// ── Validation ──────────────────────────────────────────────────────────────

// TemplateValidationResult holds validation output.
type TemplateValidationResult struct {
	Valid        bool     `json:"valid"`
	Variables    []string `json:"variables"`
	HasSpintax   bool     `json:"has_spintax"`
	Errors       []string `json:"errors"`
	Warnings     []string `json:"warnings"`
	MissingVars  []string `json:"missing_vars,omitempty"`
}

// ValidateTemplate checks a template for syntax issues.
func ValidateTemplate(template string) TemplateValidationResult {
	result := TemplateValidationResult{
		Valid:    true,
		Errors:   []string{},
		Warnings: []string{},
	}

	// Empty template check
	if strings.TrimSpace(template) == "" {
		result.Errors = append(result.Errors, "Template is empty")
		result.Valid = false
		return result
	}

	// Extract variables
	matches := templateVarRegex.FindAllStringSubmatch(template, -1)
	seen := make(map[string]bool)
	for _, m := range matches {
		if !seen[m[1]] {
			result.Variables = append(result.Variables, m[1])
			seen[m[1]] = true
		}
	}

	// Check for spintax
	result.HasSpintax = spintaxRegex.MatchString(template)

	// Validate known variables
	knownVars := map[string]bool{
		"firstName": true, "lastName": true, "fullName": true, "name": true,
		"company": true, "position": true, "jobTitle": true, "location": true,
		"headline": true, "email": true, "linkedinUrl": true, "aiIcebreaker": true,
	}

	for _, v := range result.Variables {
		if !knownVars[v] {
			result.Errors = append(result.Errors, "Unknown variable: {{"+v+"}}")
			result.Valid = false
		}
	}

	// Check for unmatched template braces
	openCount := strings.Count(template, "{{")
	closeCount := strings.Count(template, "}}")
	if openCount != closeCount {
		result.Errors = append(result.Errors, "Unmatched template braces")
		result.Valid = false
	}

	// Check for unclosed spintax braces (single { } not inside {{ }})
	// Strip out {{ }} first, then count remaining { and }
	stripped := templateVarRegex.ReplaceAllString(template, "")
	spintaxOpen := strings.Count(stripped, "{")
	spintaxClose := strings.Count(stripped, "}")
	if spintaxOpen != spintaxClose {
		result.Errors = append(result.Errors, "Unclosed spintax braces")
		result.Valid = false
	}

	// Check for nested braces (e.g. {outer {inner} text})
	depth := 0
	for _, ch := range stripped {
		if ch == '{' {
			depth++
			if depth > 1 {
				result.Errors = append(result.Errors, "Nested braces are not supported")
				result.Valid = false
				break
			}
		} else if ch == '}' {
			depth--
		}
	}

	return result
}

// GetAvailableVariables returns the list of supported template variables.
func GetAvailableVariables() []map[string]string {
	return []map[string]string{
		{"name": "firstName", "description": "Lead's first name"},
		{"name": "lastName", "description": "Lead's last name"},
		{"name": "fullName", "description": "Lead's full name"},
		{"name": "company", "description": "Lead's company"},
		{"name": "position", "description": "Lead's job title"},
		{"name": "location", "description": "Lead's location"},
		{"name": "headline", "description": "Lead's LinkedIn headline"},
		{"name": "email", "description": "Lead's email address"},
		{"name": "linkedinUrl", "description": "Lead's LinkedIn profile URL"},
		{"name": "aiIcebreaker", "description": "AI-generated icebreaker sentence"},
	}
}

// ── Additional Helpers ──────────────────────────────────────────────────────

// ExtractVariables returns the unique variable names used in a template.
func ExtractVariables(template string) []string {
	matches := templateVarRegex.FindAllStringSubmatch(template, -1)
	seen := make(map[string]bool)
	var vars []string
	for _, m := range matches {
		if len(m) >= 2 && !seen[m[1]] {
			vars = append(vars, m[1])
			seen[m[1]] = true
		}
	}
	return vars
}

// MissingDataResult holds the result of checking required template data.
type MissingDataResult struct {
	HasAllData  bool     `json:"has_all_data"`
	MissingVars []string `json:"missing_vars,omitempty"`
}

// HasRequiredData checks if a lead has all variable data needed for a template.
// Returns both a bool and the list of missing variable names.
func HasRequiredData(template string, vars *TemplateVariables) MissingDataResult {
	if vars == nil {
		needed := ExtractVariables(template)
		return MissingDataResult{HasAllData: false, MissingVars: needed}
	}
	needed := ExtractVariables(template)
	varMap := vars.variableMap()
	var missing []string
	for _, v := range needed {
		if val, ok := varMap[v]; !ok || val == "" {
			missing = append(missing, v)
		}
	}
	return MissingDataResult{
		HasAllData:  len(missing) == 0,
		MissingVars: missing,
	}
}
